src/controllers/verify-yc.ts

```diff
@@ -0,0 +1,134 @@
+import { Body, Controller, Ctx, Post } from 'amala'
+import { Context } from 'vm'
+import { badRequest } from '@hapi/boom'
+import { ethers, utils } from 'ethers'
+import { polygonProvider } from '@/helpers/providers'
+import AddressVerifyBody from '@/validators/AddressVerifyBody'
+import BalanceUniqueVerifyBody from '@/validators/BalanceUniqueVerifyBody'
+import EmailUniqueVerifyBody from '@/validators/EmailUniqueVerifyBody'
+import TwitterBody from '@/validators/TwitterBody'
+import eddsaSigPoseidon from '@/helpers/signatures/eddsaSigPoseidon'
+import fetchUserProfile from '@/helpers/twitter/fetchUserProfile'
+import getBalance from '@/helpers/getBalance'
+import poseidonHash from '@/helpers/signatures/poseidonHash'
+import sendEmail from '@/helpers/sendEmail'
+import zeroAddress from '@/models/zeroAddress'
+
+@Controller('/verify-yc')
+export default class VerifyYCController {
+  @Post('/email-unique')
+  async sendUniqueEmail(
+    @Body({ required: true }) { email }: EmailUniqueVerifyBody
+  ) {
+    // 1. Create `emailHash = poseidon([0, email])`, (0 = email attestation)
+    const eddsaMessage = ['0', utils.hexlify(utils.toUtf8Bytes(email))]
+
+    const emailHash = await poseidonHash(eddsaMessage)
+
+    // 2. Sign `[0, emailHash]` with EdDSA, (0 = yc attestation)
+    const eddsaSignature = await eddsaSigPoseidon([
+      0,
+      ...utils.toUtf8Bytes(emailHash),
+    ])
+
+    const domain = email.split('@')[1].toLowerCase()
+
+    // TODO:
+    // 3. Send email with signature (must look nice and ketl branded)
+    void sendEmail({
+      domain,
+      secret: eddsaSignature,
+      subject: "Here's your token!",
+      to: email,
+    })
+  }
+
+  @Post('/twitter')
+  async twitter(
+    @Ctx() ctx: Context,
+    @Body({ required: true }) { token }: TwitterBody
+  ) {
+    // 1. Verify token
+    let userId = ''
+    try {
+      const { id } = await fetchUserProfile(token)
+      userId = id
+    } catch (e) {
+      console.log(e)
+      return ctx.throw(badRequest('Failed to fetch user profile'))
+    }
+
+    // 2. Get userIdHash = poseidon([1, userId]), (1 = twitter attestation)
+    const eddsaMessage = ['1', userId]
+    const userIdHash = await poseidonHash(eddsaMessage)
+
+    // 3. Sign `[0, userIdHash]` with EdDSA, (0 = yc attestation)
+    const eddsaSignature = await eddsaSigPoseidon([0, userIdHash])
+
+    // 4. Return signature
+    return {
+      message: [0, userIdHash],
+      signature: eddsaSignature,
+    }
+  }
+
+  @Post('/balance-unique')
+  async sendUniqueBalance(
+    @Ctx() ctx: Context,
+    @Body({ required: true })
+    {
+      message,
+      ownerAddress,
+      signature,
+      threshold,
+      tokenAddress = zeroAddress,
+      tokenId,
+    }: BalanceUniqueVerifyBody & AddressVerifyBody
+  ) {
+    // 1. Verify signature and that ownerAddress signed the signature
```

Comment:
We should encapsulate logic into helper functions with appropriate names to get rid of those comments